cpp/src/io/fst/logical_stack.cuh

```diff
@@ -0,0 +1,530 @@
+  // Initialize double-buffer for sorting the indexes of the sequence of sparse stack operations
+  d_symbol_positions_db =
+    cub::DoubleBuffer<SymbolPositionT>{d_symbol_positions.data(), d_symbol_position_alt.data()};
+
+  // Initialize double-buffer for sorting the indexes of the sequence of sparse stack operations
+  d_kv_operations = cub::DoubleBuffer<StackOpT>{d_kv_ops_current.data(), d_kv_ops_alt.data()};
+
+  // Compute prefix sum of the stack level after each operation
+  CUDF_CUDA_TRY(cub::DeviceScan::InclusiveScan(temp_storage.data(),
+                                               total_temp_storage_bytes,
+                                               stack_symbols_in,
+                                               d_kv_operations.Current(),
+                                               detail::AddStackLevelFromStackOp{},
+                                               num_symbols_in,
+                                               stream));
+
+  // Dump info on stack operations: (stack level change + symbol) -> (absolute stack level + symbol)
+  test::print::print_array(num_symbols_in,
+                           stream,
+                           make_stack_level_iterator(stack_symbols_in),
+                           make_value_iterator(stack_symbols_in),
+                           make_stack_level_iterator(d_kv_operations.Current()),
+                           make_value_iterator(d_kv_operations.Current()));
+
+  // Stable radix sort, sorting by stack level of the operations
+  d_kv_operations_unsigned = cub::DoubleBuffer<StackOpUnsignedT>{
+    reinterpret_cast<StackOpUnsignedT*>(d_kv_operations.Current()),
+    reinterpret_cast<StackOpUnsignedT*>(d_kv_operations.Alternate())};
+  CUDF_CUDA_TRY(cub::DeviceRadixSort::SortPairs(temp_storage.data(),
+                                                total_temp_storage_bytes,
+                                                d_kv_operations_unsigned,
+                                                d_symbol_positions_db,
+                                                num_symbols_in,
+                                                begin_bit,
+                                                end_bit,
+                                                stream));
+
+  // TransformInputIterator that remaps all operations on stack level 0 to the empty stack symbol
+  kv_ops_scan_in  = {reinterpret_cast<StackOpT*>(d_kv_operations_unsigned.Current()),
+                    detail::RemapEmptyStack<StackOpT>{empty_stack}};
+  kv_ops_scan_out = reinterpret_cast<StackOpT*>(d_kv_operations_unsigned.Alternate());
+
+  // Dump info on stack operations sorted by their stack level (i.e. stack level after applying
+  // operation)
+  test::print::print_array(num_symbols_in,
+                           stream,
+                           make_stack_level_iterator(kv_ops_scan_in),
+                           make_value_iterator(kv_ops_scan_in));
```

Comment:
Do not forget to remove these debug-print.

Reply:
`print_array` would check for the value of `CUDA_DBG_DUMP` environment variable and only do more heavy-weight stuff then. Nonetheless, I've removed it to not even incur the overhead of checking the env var.